clang/lib/Analysis/LifetimeSafety/LoanPropagation.cpp

```diff
@@ -7,34 +7,100 @@
 //===----------------------------------------------------------------------===//
 #include "clang/Analysis/Analyses/LifetimeSafety/LoanPropagation.h"
 #include "Dataflow.h"
+#include "llvm/ADT/BitVector.h"
+#include "llvm/Support/TimeProfiler.h"
 #include <memory>
 
 namespace clang::lifetimes::internal {
+
+// Pre-pass to find persistent origins. An origin is persistent if it is
+// referenced in more than one basic block.
+static llvm::BitVector computePersistentOrigins(FactManager &FactMgr,
+                                                const CFG &C) {
+  llvm::TimeTraceScope("ComputePersistentOrigins");
+  llvm::BitVector PersistentOrigins(FactMgr.getOriginMgr().getOrigins().size() +
+                                    1);
+
+  llvm::DenseMap<OriginID, const CFGBlock *> OriginToBlock;
+  for (const CFGBlock *B : C) {
+    for (const Fact *F : FactMgr.getFacts(B)) {
+      auto CheckOrigin = [&](OriginID OID) {
+        if (PersistentOrigins.test(OID.Value))
+          return;
+        auto It = OriginToBlock.find(OID);
```

Comment:
Nit: we could use try_emplace to save lookup (id we don't want to switch to a vector).